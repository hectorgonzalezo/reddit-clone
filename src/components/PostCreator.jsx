import React, { useRef, useState } from 'react';
import CommunityChooser from './CommunityChooser';
import PostingRules from './PostingRules';
import ImageUpload from './ImageUpload';
import Button from './Button';
import postIcon from '../assets/post_icon.svg';
import imagesIcon from '../assets/images_icon.svg';
import linkIcon from '../assets/link_icon.svg';
import '../styles/postCreatorStyle.scss';

function PostCreator() {
  const selectedButton = useRef();
  const [mediaType, setMediaType] = useState(<textarea name="" id="" cols="30" rows="10" placeholder="Text (optional)" />)
  const textArea = <textarea id="text-area" cols="30" rows="10" placeholder="Text (optional)" />;
  const urlArea = <textarea id="url-area" cols="30" rows="2" placeholder="Url" required /> ;

   // switches between text, images and link inputs
   function addMediaType() {
    const type = selectedButton.current.getAttribute('data');
    switch (type) {
      case 'post':
        setMediaType(textArea)
        break;
      case 'image':
        setMediaType(<ImageUpload />);
        break;
      case 'link':
        setMediaType(urlArea)
        break;
      default:
        break;
    };
  }

  // Selects a type of media button
  function selectMedia(e) {
    selectedButton.current.classList.remove('selected');
    e.target.classList.add('selected');
    selectedButton.current = e.target;
    
    addMediaType();
  }

 

  return (
    <>
      <div id="left-side">
        <h1>Create a post</h1>
        <hr />
        <CommunityChooser />
        <div id="post-creator" className="main-child">
          <div id="buttons-div" data-testid="buttons-div">
            <button
              type="button"
              className="selected"
              ref={selectedButton}
              onClick={selectMedia}
              data="post"
            >
              <img src={postIcon} alt="" />
              Post
            </button>
            <button type="button" onClick={selectMedia} data="image">
              <img src={imagesIcon} alt="" />
              Images & Video
            </button>
            <button type="button" onClick={selectMedia} data="link">
              <img src={linkIcon} alt="" />
              Link
            </button>
          </div>
          <form action="" id="post-form" aria-label="Submit Form">
            <input type="text" placeholder="Title" />
            {mediaType}
            <Button text="Post" />
          </form>
        </div>
      </div>
      <div id="right-side">
       <PostingRules /> 
      </div>
    </>
  );
}

export default PostCreator;
