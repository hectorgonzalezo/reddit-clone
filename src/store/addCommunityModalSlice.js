import { createSlice } from '@reduxjs/toolkit';

export const addCommunityModalSlice = createSlice({
  name: 'addCommunityModalVisible',
  initialState: false,
  reducers: {
    toggleAddCommunityModal: (state, action) => {
      return !state
    }
  },
});

export const { toggleAddCommunityModal } = addCommunityModalSlice.actions;

export const selectAddCommunityModalVisibility = (state) => state.addCommunityModalVisible;

export default addCommunityModalSlice.reducer;
