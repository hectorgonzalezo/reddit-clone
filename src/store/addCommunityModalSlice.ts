import { createSlice } from '@reduxjs/toolkit';

export const addCommunityModalSlice = createSlice({
  name: 'addCommunityModalVisible',
  initialState: false,
  reducers: {
    toggleAddCommunityModal: (state) => !state
  },
});

export const { toggleAddCommunityModal } = addCommunityModalSlice.actions;

export const selectAddCommunityModalVisibility = (state: AddCommunityModalState): boolean => state.addCommunityModalVisible;

export default addCommunityModalSlice.reducer;
