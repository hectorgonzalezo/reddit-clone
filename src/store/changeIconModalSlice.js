import { createSlice } from '@reduxjs/toolkit';

export const changeIconModalSlice = createSlice({
  name: 'changeIconModalVisible',
  initialState: false,
  reducers: {
    toggleChangeIconModal: (state, action) => {
      return !state
    }
  },
});

export const { toggleChangeIconModal } = changeIconModalSlice.actions;

export const selectChangeIconModalVisibility = (state) => state.changeIconModalVisible;

export default changeIconModalSlice.reducer;
