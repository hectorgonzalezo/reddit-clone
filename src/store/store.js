import { configureStore } from '@reduxjs/toolkit';
import userReducer from './userSlice';
import subredditsReducer from './subredditsSlice';
import loginModalReducer from './loginModalSlice';
import currentSubredditReducer from './currentSubredditSlice';

const store = configureStore({
  reducer: {
    user: userReducer,
    subreddits: subredditsReducer,
    loginModalVisible: loginModalReducer,
    currentSubreddit: currentSubredditReducer,
  }
});

export default store;
