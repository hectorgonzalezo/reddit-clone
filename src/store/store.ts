import { configureStore } from '@reduxjs/toolkit';
import userReducer from './userSlice';
import postsReducer from './postsSlice';
import subredditsReducer from './subredditsSlice';
import loginModalReducer from './loginModalSlice';
import changeIconModalReducer from './changeIconModalSlice';
import addCommunityModalReducer from './addCommunityModalSlice';
import currentSubredditReducer from './currentSubredditSlice';

const store = configureStore({
  reducer: {
    user: userReducer,
    posts: postsReducer,
    subreddits: subredditsReducer,
    loginModalVisible: loginModalReducer,
    changeIconModalVisible: changeIconModalReducer,
    addCommunityModalVisible: addCommunityModalReducer,
    currentSubreddit: currentSubredditReducer,
  }
});

export default store;
