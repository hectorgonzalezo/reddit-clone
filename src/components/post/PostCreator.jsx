import React, { useRef, useState } from 'react';
import { useSelector } from 'react-redux';
import { useParams, useNavigate } from 'react-router-dom';
import { selectUser } from '../../store/userSlice';
import CommunityChooser from '../CommunityChooser';
import PostingRules from './PostingRules';
import ImageUpload from '../ImageUpload';
import Button from '../Button';
import { database } from '../../firebase/firebase';
import postIcon from '../../assets/post_icon.svg';
import imagesIcon from '../../assets/images_icon.svg';
import linkIcon from '../../assets/link_icon.svg';
import loadingIcon from '../../assets/loading.gif';
import '../../styles/postCreatorStyle.scss';

function PostCreator() {
  const user = useSelector(selectUser);
  const navigate = useNavigate();
  const selectedButton = useRef();
  const titleRef = useRef();
  const [textContent, setTextContent] = useState('');
  const [selectedSubreddit, setSelectedSubreddit] = useState(null);
  const [fileToUpload, setFileToUpload] = useState(null);
  const [isLoading, setIsLoading] = useState(false);
  let { initialType } = useParams();
  // by default should be textArea
  if (initialType === undefined) {
    initialType = 'textArea';
  }

  const inputs = {
    textArea: (
      <textarea
        id="text-area"
        cols="30"
        rows="10"
        placeholder="Text (optional)"
        onChange={updateTextContent}
      />
    ),
    urlArea: (
      <textarea
        id="url-area"
        cols="30"
        rows="2"
        placeholder="Url"
        onChange={updateTextContent}
        required
      />
    ),
    imgArea: <ImageUpload id="img-area" onChange={updateFileToUpload} />,
  };
  const [mediaType, setMediaType] = useState(inputs[initialType]);

  // used by both text areas onChange
  function updateTextContent(e) {
    const content = e.target.value;
    setTextContent(content);
  }

  // used by imgArea
  function updateFileToUpload(e) {
    const newFile = e.target.files[0];
    setFileToUpload(newFile);
  }

  // switches between text, images and link inputs
  function addMediaType() {
    const type = selectedButton.current.getAttribute('data');
    switch (type) {
      case 'post':
        setMediaType(inputs.textArea);
        break;
      case 'image':
        setMediaType(inputs.imgArea);
        break;
      case 'link':
        setMediaType(inputs.urlArea);
        break;
      default:
        break;
    }
  }

  // Selects a type of media button
  function selectMedia(e) {
    selectedButton.current.classList.remove('selected');
    e.target.classList.add('selected');
    selectedButton.current = e.target;
    addMediaType();
  }

  // adds a post to database
  async function submitPost() {
    const { username } = user;
    const title = titleRef.current.value;
    const subreddit = selectedSubreddit;
    const text = textContent;
    const image = fileToUpload;
    let id;

    // add loading animation to button
    setIsLoading(true);

    // depending on the type of media to be upload, choose an appropriate database method
    // extract the id of the element to identify the type
    switch (mediaType.props.id) {
      case 'text-area':
        id = await database.addTextPost(username, title, subreddit, text);
        loadPost(id);
        break;
      case 'img-area':
        id = await database.addImagePost(username, title, subreddit, image);
        loadPost(id);
        break;
      case 'url-area':
        id = await database.addUrlPost(username, title, subreddit, text);
        loadPost(id);
        break;
      default:
        setIsLoading(false);
        break;
    }
  }

  function loadPost(id) {
    setIsLoading(false);
    navigate(`/r/${selectedSubreddit}/${id}`);
  }

  return (
    <>
      <div id="left-side">
        <h1 id="create-post-title">Create a post</h1>
        <CommunityChooser onChoosing={setSelectedSubreddit} />
        <div id="post-creator" className="main-child">
          <div id="buttons-div" data-testid="buttons-div">
            <button
              type="button"
              className={initialType === 'textArea' ? "selected" : ''}
              ref={selectedButton}
              onClick={selectMedia}
              data="post"
            >
              <img src={postIcon} alt="" />
              Post
            </button>
            <button type="button" className={initialType === 'imgArea' ? "selected" : ''} onClick={selectMedia} data="image">
              <img src={imagesIcon} alt="" />
              Images & Video
            </button>
            <button type="button" className={initialType === 'urlArea' ? "selected" : ''} onClick={selectMedia} data="link">
              <img src={linkIcon} alt="" />
              Link
            </button>
          </div>
          <form action="" id="post-form" aria-label="Submit Form">
            <input type="text" placeholder="Title" ref={titleRef} />
            {mediaType}
            <Button onClick={submitPost} disabled={selectedSubreddit === null ? true : false}>
              {isLoading ? <img src={loadingIcon} alt="loading icon" className="loading-icon" /> : 'Post'}
            </Button>
          </form>
        </div>
      </div>
      <div id="right-side">
        <PostingRules />
      </div>
    </>
  );
}

export default PostCreator;
