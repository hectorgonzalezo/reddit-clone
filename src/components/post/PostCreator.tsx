import React, { SyntheticEvent, useRef, useState } from 'react';
import { useSelector } from 'react-redux';
import { useParams, useNavigate } from 'react-router-dom';
import { selectUser } from '../../store/userSlice';
import CommunityChooser from '../CommunityChooser';
import PostingRules from './PostingRules';
import ImageUpload from '../ImageUpload';
import Button from '../Button';
import { database } from '../../firebase/firebase';
import postIcon from '../../assets/post_icon.svg';
import imagesIcon from '../../assets/images_icon.svg';
import linkIcon from '../../assets/link_icon.svg';
import loadingIcon from '../../assets/loading.gif';
import '../../styles/postCreatorStyle.scss';

function PostCreator(): JSX.Element {
  const user = useSelector(selectUser);
  const navigate = useNavigate();
  const selectedButton = useRef<HTMLButtonElement | null>(null);
  const titleRef = useRef<HTMLInputElement | null>(null);
  const [textContent, setTextContent] = useState('');
  const [selectedSubreddit, setSelectedSubreddit] = useState<string>();
  const [fileToUpload, setFileToUpload] = useState<File>();
  const [isLoading, setIsLoading] = useState(false);
  let { initialType } = useParams<string>();
  // by default should be textArea
  if (initialType === undefined) {
    initialType = 'textArea';
  }

  const inputs: { [index: string]: JSX.Element } = {
    textArea: (
      <textarea
        id="text-area"
        cols={30}
        rows={10}
        placeholder="Text (optional)"
        onChange={updateTextContent}
      />
    ),
    urlArea: (
      <textarea
        id="url-area"
        cols={30}
        rows={2}
        placeholder="Url"
        onChange={updateTextContent}
        required
      />
    ),
    imgArea: (
      <ImageUpload id="img-area" onChange={updateFileToUpload} required />
    ),
  };

  const [mediaType, setMediaType] = useState<JSX.Element>(inputs[initialType]);

  // used by both text areas onChange
  function updateTextContent(e: SyntheticEvent): void {
    const target = e.target as HTMLInputElement;
    const content = target.value;
    setTextContent(content);
  }

  // used by imgArea
  function updateFileToUpload(e: SyntheticEvent): void {
    const target = e.target as HTMLInputElement;
    if (target.files) {
      const newFile = target.files[0];
      setFileToUpload(newFile);
    }
  }

  // switches between text, images and link inputs
  function addMediaType(): void {
    if (selectedButton.current !== null) {
    const type = selectedButton.current.getAttribute('data-type');
    switch (type) {
      case 'post':
        setMediaType(inputs.textArea);
        break;
      case 'image':
        setMediaType(inputs.imgArea);
        break;
      case 'link':
        setMediaType(inputs.urlArea);
        break;
      default:
        break;
    }
  }
  }

  // Selects a type of media button
  function selectMedia(e: SyntheticEvent): void {
    const target = e.target as HTMLButtonElement;
    if (selectedButton.current !== null) {
      selectedButton.current.classList.remove('selected');
      target.classList.add('selected');
      selectedButton.current = target;
      addMediaType();
    }
  }

  // adds a post to database
  async function submitPost(): Promise<void> {
    const { username } = user;
    if ( titleRef.current !== null) {
    const title = titleRef.current.value;
    const subreddit = selectedSubreddit;
    const text = textContent;
    const image = fileToUpload;
    let id;

    // add loading animation to button
    setIsLoading(true);

    // depending on the type of media to be upload, choose an appropriate database method
    // extract the id of the element to identify the type
    switch (mediaType.props.id) {
      case 'text-area':
        id = await database.addTextPost(username, title, subreddit, text);
        loadPost(id);
        break;
      case 'img-area':
        id = await database.addImagePost(username, title, subreddit, image);
        loadPost(id);
        break;
      case 'url-area':
        id = await database.addUrlPost(username, title, subreddit, text);
        loadPost(id);
        break;
      default:
        setIsLoading(false);
        break;
    }
  }
  }

  function loadPost(id: string): void {
    setTimeout(() => {
      setIsLoading(false);
      navigate(`/r/${selectedSubreddit as string}/${id}`);
    }, 1000);
  }

  return (
    <>
      <div id="left-side">
        <h1 id="create-post-title">Create a post</h1>
        <CommunityChooser onChoosing={setSelectedSubreddit} />
        <div id="post-creator" className="main-child">
          <div id="buttons-div" data-testid="buttons-div">
            <button
              type="button"
              className={initialType === "textArea" ? "selected" : ""}
              ref={selectedButton}
              onClick={selectMedia}
              data-type="post"
            >
              <img src={postIcon} alt="" />
              Post
            </button>
            <button
              type="button"
              className={initialType === "imgArea" ? "selected" : ""}
              onClick={selectMedia}
              data-type="image"
            >
              <img src={imagesIcon} alt="" />
              Images & Video
            </button>
            <button
              type="button"
              className={initialType === "urlArea" ? "selected" : ""}
              onClick={selectMedia}
              data-type="link"
            >
              <img src={linkIcon} alt="" />
              Link
            </button>
          </div>
          <form action="" id="post-form" aria-label="Submit Form">
            <input type="text" placeholder="Title" ref={titleRef} />
            {mediaType}
            <Button
              onClick={submitPost}
              disabled={selectedSubreddit === null ? true : false}
            >
              {isLoading ? (
                <img
                  src={loadingIcon}
                  alt="loading icon"
                  className="loading-icon"
                />
              ) : (
                "Post"
              )}
            </Button>
          </form>
        </div>
      </div>
      <div id="right-side">
        <PostingRules />
      </div>
    </>
  );
}

export default PostCreator;
