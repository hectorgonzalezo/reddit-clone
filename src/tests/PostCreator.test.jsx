import React from 'react';
import { Provider } from 'react-redux';
import { render, screen, container, act, getByRole, queryAllByRole } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import store from '../store/store';
import PostCreator from '../components/post/PostCreator';


describe('Post creator page', () => {
  test('Post starts highlighted', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <PostCreator />
        </BrowserRouter>
      </Provider>
    );
    const buttonsArea = screen.getByTestId('buttons-div');
    const postButton = getByRole(buttonsArea, 'button', { name: 'Post' });
    
    expect(postButton).toHaveClass('selected');

    expect(container).toMatchSnapshot();
  });

  test('Pressing on media button highlights it and deselects the rest', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <PostCreator />
        </BrowserRouter>
      </Provider>
    );
    const buttonsArea = screen.getByTestId('buttons-div');;
    const postButton = getByRole(buttonsArea, 'button', { name: 'Post' });
    const imagesButton = getByRole(buttonsArea, 'button', { name: 'Images & Video' });
    const linkButton = getByRole(buttonsArea, 'button', { name: 'Link' });

    userEvent.click(linkButton);
    
    expect(postButton).not.toHaveClass('selected');
    expect(imagesButton).not.toHaveClass('selected');
    expect(linkButton).toHaveClass('selected');
  });

  test('By default shows text area with text placeholder', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <PostCreator />
        </BrowserRouter>
      </Provider>
    );
    const form = screen.getByRole('form');
    const textArea = queryAllByRole(form, 'textbox');
    
    expect(textArea[1]).toBeInTheDocument();
    expect(textArea[1].placeholder).toEqual('Text (optional)');
  });

  test('Pressing on images button shows file input', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <PostCreator />
        </BrowserRouter>
      </Provider>
    );
    const form = screen.getByRole('form');
    const buttonsArea = screen.getByTestId('buttons-div');;
    const imagesButton = getByRole(buttonsArea, 'button', { name: 'Images & Video' });

    userEvent.click(imagesButton);
    const imageInput = screen.getByText('Choose a file');
    
    expect(imageInput).toBeInTheDocument();
  });

  test('Pressing on link button shows link textarea', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <PostCreator />
        </BrowserRouter>
      </Provider>
    );
    const form = screen.getByRole('form');
    const buttonsArea = screen.getByTestId('buttons-div');;
    const linkButton = getByRole(buttonsArea, 'button', { name: 'Link' });

    userEvent.click(linkButton);
    const textArea = queryAllByRole(form, 'textbox');
    
    expect(textArea[1]).toBeInTheDocument();
    expect(textArea[1].placeholder).toEqual('Url');
    expect(textArea[1]).toHaveProperty('required', true);
  });

  test('Pressing on "choose a community" shows drop down', async () => {

    render(
      <Provider store={store}>
        <BrowserRouter>
          <PostCreator />
        </BrowserRouter>
      </Provider>
    );

    const dropDown = screen.getByTestId('subreddits-dropdown');
    const button = screen.getByText('Choose a community');

    expect(dropDown).toHaveStyle('display: none');

    userEvent.click(button);

    expect(dropDown).toHaveStyle('display: flex');
  });

  test('Choosing a subreddit displays its icon and name', async () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <PostCreator />
        </BrowserRouter>
      </Provider>
    );

    await act(async () => {
      store.dispatch({
        type: "subreddits/addSubreddit",
        payload: [{
          name: "aww",
          icon: "https://firebasestorage.googleapis.com/v0/b/reddit-clone-83ce9.appspot.com/o/aww_icon.jpeg?alt=media&token=b6bc425f-f335-4b73-8c36-df3df45bacd6",
        }],
      });
    });

    const dropDown = screen.getByTestId('subreddits-dropdown');
    const button = screen.getByText('Choose a community');
    userEvent.click(button);

    const subreditLink = getByRole(dropDown, 'img');
    userEvent.click(subreditLink);

    const buttonTitle = screen.getByTestId('community-chooser-title');

    expect(buttonTitle).toHaveTextContent('aww');
  });
})