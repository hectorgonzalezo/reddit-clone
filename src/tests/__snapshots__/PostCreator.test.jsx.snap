// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post creator page Post starts highlighted 1`] = `undefined`;
